backend/auth/role_map_repository.go

```diff
@@ -0,0 +1,279 @@
+package auth
+
+import (
+	"errors"
+	"log"
+	"os"
+	"sync"
+	"github.com/ZPI-2024-25/KubernetesAccessManager/cluster"
+	"github.com/ZPI-2024-25/KubernetesAccessManager/models"
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type RoleMapRepository struct {
+	// Subrole map is optional, if not provided, subroles will be ignored. Roles received in token are checked only with RoleMap
+	RoleMap map[string]*models.Role
+	SubroleMap map[string]*models.Role
+	flattenedMap map[string]map[string]map[string][]models.OperationType
+}
+
+var (
+	instance      *RoleMapRepository
+	once          sync.Once
+)
+
+func GetRoleMapInstance() (*RoleMapRepository, error) {
+	once.Do(func() {
+		roleMapNamespace := os.Getenv("ROLEMAP_NAMESPACE")
+		if roleMapNamespace == "" {
+			log.Printf("ROLEMAP_NAMESPACE not set, using default namespace")
+			roleMapNamespace = "default"
+		}
+
+		roleMapName := os.Getenv("ROLEMAP_NAME")
+		if roleMapName == "" {
+			log.Printf("ROLEMAP_NAME not set, using default name")
+			roleMapName = "role-mapper"
+		}
+
+		roleMap, subroleMap := GetRoleMapConfig(roleMapNamespace, roleMapName)
+		if roleMap == nil {
+			log.Printf("Failed to initialize RoleMapRepository")
+			return
+		}
+		instance = &RoleMapRepository{
+			RoleMap: roleMap, 
+			SubroleMap: subroleMap,
+		}
+
+		log.Printf("RoleMapRepository initialized with %d roles %d subroles", len(instance.RoleMap), len(instance.SubroleMap))
+	})
+
+	if instance == nil {
+		return nil, errors.New("failed to initialize RoleMapRepository")
+	}
+
+	return instance, nil
+}
+
+func (rmr *RoleMapRepository) HasPermission(rolenames []string, operation *models.Operation) bool {
```

Comment:
This will change when we implement the role matrix?

Reply:
The definition, the implementation yes